src/container/srv_container.c

```diff
@@ -369,55 +369,180 @@ enum nhandles_op {
+{
+	int rc;
+
+	rc = d_hash_table_destroy_inplace(nht, true);
+	if (rc)
+		D_ERROR("d_hash_table_destroy_inplace() failed, "DF_RC"\n", DP_RC(rc));
+	return rc;
+}
+
+/* Get, optionally update container number of handles. Use "nhandles cache" (nhc) if provided. */
+static int
+get_nhandles(struct rdb_tx *tx, struct d_hash_table *nhc, struct cont *cont, enum nhandles_op op,
+	     uint32_t *nhandles)
+{
+	uint32_t		result = 0;
+	d_iov_t			value;
+	uint32_t		lookup_val;		/* value from DRAM cache or rdb */
+	struct nhandles_ht_rec *rec = NULL;
+	struct nhandles_ht_rec *rec_inserted = NULL;
+	bool			do_update = true;
+	uint32_t		pool_global_version = cont->c_svc->cs_pool->sp_global_version;
+	int			rc;
+
+	/* Test if pool/container has old layout without nhandles */
+	if (pool_global_version < DAOS_POOL_GLOBAL_VERSION_WITH_CONT_NHANDLES)
+		goto out;
+
+	/* Caller performing multiple updates in the tx: insert into, or get value from HT cache. */
+	if (nhc) {
+		d_list_t *hlink = d_hash_rec_find(nhc, cont->c_uuid, sizeof(uuid_t));
+
+		if (hlink) {
+			rec = nhandles_rec_obj(hlink);
+			lookup_val = rec->nhr_nhandles;
+			d_hash_rec_decref(nhc, &rec->nhr_hlink);
+		}
+	}
+
+	/* If no cache, or miss, lookup value in rdb. */
+	if (rec == NULL) {
+		d_iov_set(&value, &lookup_val, sizeof(lookup_val));
+		rc = rdb_tx_lookup(tx, &cont->c_prop, &ds_cont_prop_nhandles, &value);
+		if (rc) {
+			D_ERROR(DF_CONT": rdb_tx_lookup nhandles failed, "DF_RC"\n",
+				DP_CONT(cont->c_svc->cs_pool_uuid, cont->c_uuid), DP_RC(rc));
+			goto err;
+		}
+
+		/* ... and cache it, if applicable. */
+		if (nhc) {
+			D_ALLOC_PTR(rec);
+			if (rec == NULL)
+				D_GOTO(err, rc = -DER_NOMEM);
+			D_DEBUG(DB_MD, DF_CONT": alloc rec=%p\n",
+				DP_CONT(cont->c_svc->cs_pool_uuid, cont->c_uuid), rec);
+			uuid_copy(rec->nhr_cuuid, cont->c_uuid);
+			rec->nhr_nhandles = lookup_val;
+			rc = d_hash_rec_insert(nhc, rec->nhr_cuuid, sizeof(uuid_t), &rec->nhr_hlink,
+					       true);
+			if (rc != 0) {
+				D_ERROR(DF_CONT": error inserting into nhandles cache" DF_RC"\n",
+					DP_CONT(cont->c_svc->cs_pool_uuid, cont->c_uuid),
+					DP_RC(rc));
+				D_FREE(rec);
+				goto err;
+			}
+			rec_inserted = rec;
+		}
 	}
 
 	switch (op) {
 	case NHANDLES_GET:
 		do_update = false;
+		result = lookup_val;
 		break;
 	case NHANDLES_PRE_INCREMENT:
-		result++;
+		result = lookup_val + 1;
 		break;
 	case NHANDLES_PRE_DECREMENT:
-		result--;
+		result = lookup_val - 1;
 		break;
 	default:
 		D_ASSERTF(0, "invalid op=%d\n", op);
 		break;
 	}
 
+	/* Update persistent and, if applicable, cached nhandles value */
 	if (do_update) {
+		d_iov_set(&value, &result, sizeof(result));
 		rc = rdb_tx_update(tx, &cont->c_prop, &ds_cont_prop_nhandles, &value);
 		if (rc != 0) {
 			D_ERROR(DF_CONT": rdb_tx_update nhandles failed, "DF_RC"\n",
 				DP_CONT(cont->c_svc->cs_pool_uuid, cont->c_uuid), DP_RC(rc));
 			goto err;
 		}
+
+		if (rec)
+			rec->nhr_nhandles = result;
```

Comment:
[Question] Thanks for making the change, Ken. If `nhc != NULL` and we found a cache record, we'll have released the `rec` reference on new line 473 before reaching this line. The `rec` pointer is technically dangling (although we may reason that the hash table is private to the current ULT such that nobody else will free the record). Would it be better to release the `rec` reference later?

It would make the `rec` check on new line 478 _look_ safer too (although it's already safe).

When `rec` is newly allocated, we may want to set its reference count to 1, then insert it into the hash table, so that regardless of whether `rec` is pre-existing or newly-allocated, we will `decref` it.